wifi: Improve wifi driver docs (#1402)
nrfconnect/sdk-nrfxlib: nrf_wifi/README.rst

nRF70 low-level API documentation

> *********************************

Reviewer: Page title requires ###, then h2 ***, h3 ===, and so on.

.. contents::
   :local:
   :depth: 2

The nRF70 Series Wi-Fi driver provides a low-level API for use cases where the application needs to access the nRF70 Series device directly.
This is typically intended for customers who want to use the nRF70 Series device on a different platform than Zephyr.

Layers overview
===============

The nRF70 Series Wi-Fi driver consists of the following key layers:

> - **FMAC** (Firmware Media Access Control) layer is the public interface to the nRF70 Series Wi-Fi driver.

Reviewer 2: ```suggestion
- **UMAC IF** (Upper Media Access Control Interface) layer is the public interface to the nRF70 Series Wi-Fi driver.
```

  It abstracts the access to the **UMAC** (Upper Media Access Control) firmware in the nRF70 Series Wi-Fi chipset.

- **HAL** (Hardware Abstraction Layer) layer abstracts the access to the nRF70 Series Wi-Fi chipset hardware.

>   It provides a set of functions that the **FMAC** layer uses to access the hardware.

Reviewer 2: ```suggestion
  It provides a set of functions that the **UMAC IF** layer uses to access the hardware.
```

> - **BAL** (Bus Abstraction Layer) layer abstracts the access to the bus level communication to nRF70 Series Wi-Fi chipset

Reviewer 2: ```suggestion
- **Bus IF** (Bus Interface) layer abstracts the access to the bus level communication to nRF70 Series Wi-Fi chipset
```

=================================

The nRF70 Series Wi-Fi chipset consists of the following key layers:

Firmware layers
---------------

Only the key firmware layers that are relevant for the nRF70 Series Wi-Fi driver are listed here.

- **UMAC** (Upper Media Access Control) firmware layer implements the IEEE 802.11 MAC layer.
  It is responsible for managing the Wi-Fi STA MLME.

- **LMAC** (Lower Media Access Control) firmware layer implements the IEEE 802.11 MAC layer.
  It is responsible for managing all real-time operations of the Wi-Fi MAC layer.
  The **BB** (Baseband) and **RF** (Radio Frequency) layers are part of the **LMAC** firmware.
  The nRF70 Series Wi-Fi driver does not interact with the **LMAC** firmware directly.

Hardware blocks
---------------

Only the key hardware blocks that are relevant for the nRF70 Series Wi-Fi driver are listed here.

- **HPQM** (Host Processor Queue Manager) hardware block is responsible for managing the control flow by providing a set of queues with addresses of commands and event buffers.
  The **HPQM** block is used by the **UMAC** firmware to communicate with the host and vice versa.

  **HPQM** queues:

>    - ``cmd_avl_queue``: This queue is used by the **UMAC** firmware to notify the host that a command processing is done and the command buffer can be reused.

Reviewer 2: ```suggestion
   - ``cmd_avl_queue``: This queue is used by the **UMAC** firmware to notify the host about command buffers that are available for the host to send commands to the firmware.
```

   - ``cmd_busy_queue``: This queue is used by the host to send commands to the **UMAC** firmware.
   - ``event_busy_queue``: This queue is used by the **UMAC** firmware to send events to the host.

>    - ``event_avl_queue``: This queue is used by the host to notify the **UMAC** firmware that an event  processing is done and the event buffer can be reused.

Reviewer 2: ```suggestion
   - ``event_avl_queue``: This queue is used by the host to notify the **UMAC** firmware that an event processing is done and the event buffer can be reused.
```

   The queue depth is controlled by the **UMAC** firmware and is not configurable by the host.
   Default queue depth is ``10`` for all queues.

Memory blocks
-------------

Only the key memory blocks that are relevant for the nRF70 Series Wi-Fi driver are listed here.

> - **Packet RAM** memory block is used to store the Transmit and Receive frames.

Reviewer 2: ```suggestion
- **Packet RAM** memory block is used to store the Transmit and Receive frames as well as the Command and Event buffers.
```

  The size of the Packet RAM is fixed (``~192KB``) and is shared between the **UMAC** firmware and the host.
  The division of the memory between the Transmit and Receive paths is configurable and is controlled by the nRF70 Series Wi-Fi driver.

Control flow: Command
=====================

A typical control flow for a **Command** for the nRF70 Series Wi-Fi driver is as follows:

> 1. **FMAC** prepares a **UMAC** command and sends it to the **HAL** layer.

Reviewer 2: ```suggestion
1. **UMAC IF** prepares a **UMAC** command and sends it to the **HAL** layer.
```

> #. **HAL** polls the **HPQM** queue for the command buffer with a timeout, in this case ``cmd_avl_queue``.

Reviewer 2: ```suggestion
#. **HAL** polls the **HPQM** queue (in this case ``cmd_avl_queue``) for a command buffer address with a timeout, .
```

> #. **HAL** dequeues the command buffer addresses from the **HPQM** ``cmd_avl_queue``.

Reviewer 2: ```suggestion
#. **HAL** dequeues the command buffer address from the **HPQM** ``cmd_avl_queue``.
```

> #. **HAL** writes the command buffer to address provided in **Step 3**.

Reviewer 2: ```suggestion
#. **HAL** writes the contents of the command to the buffer address provided in **Step 3**.
```

> #. **HAL** updates to **UMAC** that a the command is written by writing the command buffer address to the **HPQM** ``cmd_busy_queue``.

Reviewer 2: ```suggestion
#. **HAL** notifies the **UMAC** that a command is written by writing the command buffer address to the **HPQM** ``cmd_busy_queue``.
```

> #. **HAL** notifies the **UMAC** firmware that the command buffer is ready for consumption by raising  an interrupt to the **UMAC** processor.

Reviewer 2: ```suggestion
#. **HAL** notifies the **UMAC** firmware that the command buffer is ready for consumption by raising an interrupt to the **UMAC** processor.
```

Control flow: Event
===================

A typical control flow for an **Event** for the nRF70 Series Wi-Fi driver is as follows:

1. **UMAC** firmware generates an interrupt to the host processor.
#. **HAL** reads the event buffer address from the **HPQM** ``event_busy_queue``.
#. **HAL** reads the event buffer from the address provided in **Step 2**.
#. **HAL** writes the event buffer address to the **HPQM** ``event_avl_queue``.

> #. **HAL** passes the event buffer to the **FMAC** layer.

Reviewer 2: ```suggestion
#. **HAL** passes the event buffer to the **UMAC IF** layer.
```

> #. **FMAC** processes the event and depending on the event type, handles it accordingly.

Reviewer 2: ```suggestion
#. **UMAC IF** processes the event and depending on the event type, handles it accordingly.
```

.. note::

   The command and event flow is a simplified version of the actual flow.
   Fragmented commands and events are not covered in this flow.

Data flow: Transmit
===================

> **FMAC** implements a basic token bucket algorithm to flowcontrol TX frames to the nRF70 Series Wi-Fi chipset.

Reviewer 2: ```suggestion
**UMAC IF** implements a basic token bucket algorithm to flowcontrol TX frames to the nRF70 Series Wi-Fi chipset.
```

  The remaining tokens after dividing the tokens for per-AC buckets are added to the global bucket.

For example, if :kconfig:option:`CONFIG_NRF700X_MAX_TX_TOKENS` is set to ``12``, the driver divides the tokens as follows:

* Two tokens for each AC
* Two tokens for the global bucket

Queues
------

The nRF70 Series Wi-Fi driver implements the following queues:

* **TX queue** is used to store the frames that are ready to be transmitted.
  It maps to frames that can be transmitted in a single TXOP (Transmission Opportunity), for example a single A-MPDU.
  The depth of this queue is controlled by the :kconfig:option:`CONFIG_NRF700X_MAX_TX_AGGREGATION` Kconfig option and also limited by the maximum size of memory for a single token.

* **TX pending queue** is used to store the frames that are waiting for the tokens to be available in the buckets.
   The depth of this queue is controlled by the :kconfig:option:`CONFIG_NRF700X_MAX_TX_PENDING_QLEN` Kconfig option.

Flow
----

The nRF70 Series Wi-Fi driver implements the following flow for transmitting a frame:

> 1. **FMAC** receives a frame from the networking stack.

Reviewer 2: ```suggestion
1. **UMAC IF** receives a frame from the networking stack.
```

> #. **FMAC** fetches the peer, TID and AC for the frame.

Reviewer 2: ```suggestion
#. **UMAC IF** fetches the peer, TID and AC for the frame.
```

>   Depending on these the transmit pending queue is selected and the frame is enqueued.

Reviewer 3: ```suggestion
   Depending on these the transmit pending queue is selected and the frame is enqueued.
```

> #. **FMAC** checks if the frame can be transmitted immediately, by checking

Reviewer 2: ```suggestion
#. **UMAC IF** checks if the frame can be transmitted immediately, by checking
```

>    * If the frame meets the aggregation criterion, i.e. the maximum aggregation count is reached or maximum aggregation size is reached or frame cannot be aggregated.
>    * If the frame meets the token availability criterion, i.e. the frame can be transmitted based on the following:

Reviewer 3: ```suggestion
   * If the frame meets the aggregation criterion, which means the maximum aggregation count is reached or maximum aggregation size is reached or frame cannot be aggregated.
   * If the frame meets the token availability criterion, which means the frame can be transmitted based on the following:
```
We do not use the Latin-based abbreviations, i.e. is one of them.

     - Availability of the per-AC and global tokens.

> #. If the frame can be transmitted, **FMAC** dequeues the frame from the transmit pending queue and enqueues it to the transmit queue.

Reviewer 2: ```suggestion
#. If the frame can be transmitted, **UMAC IF** dequeues the frame from the transmit pending queue and enqueues it to the transmit queue.
```

> #. **FMAC** processes the transmit queue:

Reviewer 2: ```suggestion
#. **UMAC IF** processes the transmit queue:
```

   * It prepares the necessary information for **UMAC** firmware in the form of a ``struct nrf_wifi_tx_buff`` command.
   * It `copies` (this can be a memory copy or a DMA depending on underlying **BAL** implementation) each frame to the appropriate memory location in the Packet RAM

> #. **FMAC** sends the ``struct nrf_wifi_tx_buff`` command to the **HAL** layer using the typical command flow, as described in the previous section.

Reviewer 2: ```suggestion
#. **UMAC IF** sends the ``struct nrf_wifi_tx_buff`` command to the **HAL** layer using the typical command flow, as described in the previous section.
```

Data flow: Receive
==================

> The key difference between the transmit and receive flow needs to be pre-programmed in the nRF70 Series Wi-Fi chipset as packets can be received at any time.

Reviewer 2: ```suggestion
The key difference between the transmit and receive flow is that the RX packet buffers need to be pre-programmed to the nRF70 Series Wi-Fi chipset as packets can be received at any time.
```

Unlike the transmit flow, the receive flow does not have a token bucket mechanism to control the flow of packets.

Queues
------

The nRF70 Series Wi-Fi driver does not implement any queues for the receive path.
It is a circular buffer with a configurable size :kconfig:option:`CONFIG_NRF700X_RX_NUM_BUFS`.

Flow: initialization
--------------------

The nRF70 Series Wi-Fi driver implements the following flow during initialization:

1. **FMAC** allocates a set of buffers in the Packet RAM for the receive path.
    The number of buffers is controlled by the :kconfig:option:`CONFIG_NRF700X_RX_NUM_BUFS` Kconfig option.
#. **FMAC** prepares a ``struct nrf_wifi_rx_buff`` command and sends it to the **HAL** layer.
#. **HAL** sends the ``struct nrf_wifi_rx_buff`` command to the **UMAC** firmware using the typical command flow, as described in the previous section.

Flow: Receive
-------------

The nRF70 Series Wi-Fi driver implements the following flow for receiving a frame:

1. **UMAC** firmware receives a frame from the air, processes it and stores it in the Packet RAM.
    Prepares ``struct nrf_wifi_rx_buff`` and generates an interrupt to the host processor.

> #. **FMAC** receives the interrupt, processes ``struct nrf_wifi_rx_buff`` event, and copies the frame from the Packet RAM to the receive buffer.

Reviewer 2: ```suggestion
#. **UMAC IF** receives the interrupt, processes ``struct nrf_wifi_rx_buff`` event, and copies the frame from the Packet RAM to the receive buffer.
```

> #. **FMAC** performs 802.11 decapsulation and passes the frame to the networking stack.

Reviewer 2: ```suggestion
#. **UMAC IF** performs 802.11 decapsulation and passes the frame to the networking stack.
```

Flow: Refill
-------------

The nRF70 Series Wi-Fi driver implements a refill mechanism to ensure that the Packet RAM is always filled with receive buffers.

> 1. **FMAC** allocates a new buffer for the received frame.

Reviewer 2: ```suggestion
1. **UMAC IF** allocates a new buffer for the received frame.
```

> #. **FMAC** prepares a ``struct nrf_wifi_rx_buff`` command and sends it to the **HAL** layer.

Reviewer 2: ```suggestion
#. **UMAC IF** prepares a ``struct nrf_wifi_rx_buff`` command and sends it to the **HAL** layer.
```

#. **HAL** sends the ``struct nrf_wifi_rx_buff`` command to the **UMAC** firmware using the typical command flow, as described in the previous section.

.. toctree::
   :maxdepth: 1

>    :caption: Modules:

Reviewer: We cannot have a page with both the content divided in sections and a toctree of pages at the bottom. This will mess up how the hierarchy is displayed in the documentation and lead to a situation where you cannot distinguish between sections and subpages. All the new content on this page needs to be moved to a new, separate page.

The nRF70 Series Wi-Fi driver provides a low-level API for use cases where the application needs to access the nRF70 Series device directly.
This is typically intended for customers who want to use the nRF70 Series device on a different platform than Zephyr.

.. contents::
   :local:

>    :depth: 2

Reviewer: ``.. contents`` block must be placed immediately after the page title header, so before line 6.

Layers overview
===============
The nRF70 Series Wi-Fi driver consists of the following key layers:

- **FMAC** (Firmware Media Access Control) layer: This layer is the public interface to the nRF70 Series Wi-Fi driver.
  It abstracts the access to the **UMAC** (Upper Media Access Control) firmware in the nRF70 Series Wi-Fi chipset.

- **HAL** (Hardware Abstraction Layer) layer: This layer abstracts the access to the nRF70 Series Wi-Fi chipset hardware.
  It provides a set of functions that the **FMAC** layer uses to access the hardware.

- **BAL** (Bus Abstraction Layer) layer: This layer abstracts the access to the bus level communication to nRF70 Series Wi-Fi chipset
  It provides a set of functions that the **HAL** layer uses to access the bus, (Q)SPI and PCIe BUSes are supported.

nRF70 Series Wi-Fi chipset layers
=================================
The nRF70 Series Wi-Fi chipset consists of the following key layers:

> Firmware layers: Only the key firmware layers are listed here that are relevant for the nRF70 Series Wi-Fi driver.

Reviewer: Consider making this line and line 40 into a section header:
```
Firmware layers
---------------

Only the key firmware layers are listed here that are relevant for the nRF70 Series Wi-Fi driver.
```

- **UMAC** (Upper Media Access Control) firmware: This firmware layer implements the IEEE 802.11 MAC layer.
  It is responsible for managing the Wi-Fi STA MLME.

- **LMAC** (Lower Media Access Control) firmware: This firmware layer implements the IEEE 802.11 MAC layer.
  It is responsible for managing all real time operations of the Wi-Fi MAC layer.
  The **BB** (Baseband) and **RF** (Radio Frequency) layers are part of the **LMAC** firmware.
  The nRF70 Series Wi-Fi driver does not interact with the **LMAC** firmware directly.

> Hardware blocks: Only the key hardware blocks are listed here that are relevant for the nRF70 Series Wi-Fi driver.

Reviewer: Consider making this line and line 30 into a section header:
```
Hardware blocks
---------------

Only the key hardware blocks are listed here that are relevant for the nRF70 Series Wi-Fi driver.
```

- **HPQM** (Host Processor Queue Manager) hardware: This hardware block is responsible for managing the control flow by providing a set of queues with addresses of commands and event buffers.
  The **HPQM** block is used by the **UMAC** firmware to communicate with the host and vice versa.

  **HPQM** queues:
   - **cmd_avl_queue**: This queue is used by the UMAC firmware to notify the host that a command processing is done and the command buffer can be reused.
   - **cmd_busy_queue**: This queue is used by the host to send commands to the **UMAC** firmware.
   - **event_busy_queue**: This queue is used by the **UMAC** firmware to send events to the host.
   - **event_avl_queue**: This queue is used by the host to notify the **UMAC** firmware that an event  processing is done and the event buffer can be reused.

   The queue depth is controlled by the **UMAC** firmware and is not configurable by the host. Default queue depth is ``10`` for all queues.

> Memory blocks: Only the key memory blocks are listed here that are relevant for the nRF70 Series Wi-Fi driver.

Reviewer: As before, a section header?

4. **HAL** writes the command buffer to address provided in #3.
5. **HAL** updates to **UMAC** that a the command is written by writing the command buffer address to the **HPQM** queue **cmd_busy_queue**.
6. **HAL** notifies the **UMAC** firmware that the command buffer is ready for consumption by raising  an interrupt to the **UMAC** processor.

Control flow: Event
===================
A typical control flow for an **Event** for the nRF70 Series Wi-Fi driver is as follows:

1. **UMAC** firmware generates an interrupt to the host processor.
2. **HAL** reads the event buffer address from the **HPQM** queue **event_busy_queue**.
3. **HAL** reads the event buffer from the address provided in #2.
4. **HAL** writes the event buffer address to the **HPQM** queue **event_avl_queue**.
5. **HAL** passes the event buffer to the **FMAC** layer.
6. **FMAC** processes the event and depending on the event type handles it accordingly.

.. note::

   The command and event flow above is a simplified version of the actual flow, fragmented commands and events are not covered in this flow.

Data flow: Transmit
===================

> **FMAC** implements a basic token bucket algorithm to implement flowcontrol to the nRF70 Series Wi-Fi chipset.

Author: Flowcontrol is implemented in driver, so, modified:

`**FMAC** implements a basic token bucket algorithm to flowcontrol TX frames to the nRF70 Series Wi-Fi chipset.`

- 2 tokens for each AC
- 2 tokens for the global bucket

Queues
------
The nRF70 Series Wi-Fi driver implements the following queues:

- **TX queue**: This queue is used to store the frames that are ready to be transmitted.
  This queues maps to frames that can be transmitted in a single TXOP (Transmission Opportunity), e.g. a single A-MPDU.
  The depth of this queue is controlled by the :kconfig:option:`CONFIG_NRF700X_MAX_TX_AGGREGATION` Kconfig option and also limited by the maximum size of memory for a single token.
- **TX pending queue**: This queue is used to store the frames that are waiting for the tokens to be available in the buckets.
   The depth of this queue is controlled by the :kconfig:option:`CONFIG_NRF700X_MAX_TX_PENDING_QLEN` Kconfig option.

Flow
----
The nRF70 Series Wi-Fi driver implements the following flow for transmitting a frame:

1. **FMAC** receives a frame from the networking stack.
2. **FMAC** fetches the peer, TID and AC for the frame, depending on these the transmit pending queue is selected and the frame is enqueued.
3. **FMAC** checks if the frame can be transmitted immediately, by checking

>    - If the frame meets the aggregation criterion, i.e. the maximum aggregation count is reached or maximum aggregation size is reached or frame cannot be aggregated.
>    - If the frame meets the token availability criterion, i.e. the frame can be transmitted based on the following:
>      - Availability of the per-AC and global tokens.

Reviewer 3: ```suggestion
   * whether the frame meets the aggregation criterion, that is the maximum aggregation count is reached or maximum aggregation size is reached or frame cannot be aggregated.
   * whether the frame meets the token availability criterion, that is the frame can be transmitted based on the availability of the per-AC and global tokens.
```

4. If the frame can be transmitted, **FMAC** dequeues the frame from the transmit pending queue and enqueues it to the transmit queue.
5. **FMAC** processes the transmit queue:

>    - prepares necessry information for **UMAC** firmware in the form of `CMD_TX` command.
>    - `copies` (this can be a memory copy or a DMA depending on underlying **BAL** implementation) each frame to the appropriate memory location in the Packet RAM

Reviewer 3: ```suggestion
   * It prepares the necessary information for **UMAC** firmware in the form of a ``CMD_TX`` command.
   * It copies (this can be a memory copy or a DMA depending on underlying **BAL** implementation) each frame to the appropriate memory location in the Packet RAM.
```

> 6. **FMAC** sends the `CMD_TX` command to the **HAL** layer using the typical command flow, as described in the previous section.

Reviewer 3: ```suggestion
6. **FMAC** sends the ``CMD_TX`` command to the **HAL** layer using the typical command flow, as described in the previous section.
```

Data flow: Receive
==================

> Key difference between the transmit and receive flow needs to be pre-programmed in the nRF70 Series Wi-Fi chipset as packets can be received at any time.

Reviewer 3: ```suggestion
The key difference between the transmit and receive flow needs to be pre-programmed in the nRF70 Series Wi-Fi chipset as packets can be received at any time.
```

Unlike the transmit flow, the receive flow does not have a token bucket mechanism to control the flow of packets.

Queues
------

> The nRF70 Series Wi-Fi driver does not implement any queues for the receive path, its a circular buffer with a configurable size :kconfig:option:`CONFIG_NRF700X_RX_NUM_BUFS`.

Reviewer 3: ```suggestion
The nRF70 Series Wi-Fi driver does not implement any queues for the receive path.
It is a circular buffer with a configurable size :kconfig:option:`CONFIG_NRF700X_RX_NUM_BUFS`.
```

Flow: initialization
---------------------
The nRF70 Series Wi-Fi driver implements the following flow during initialization:

> 1. **FMAC** during initialization allocates a set of buffers in the Packet RAM for the receive path, the number of buffers is controlled by the :kconfig:option:`CONFIG_NRF700X_RX_NUM_BUFS` Kconfig option.

Reviewer 3: ```suggestion
1. **FMAC** allocates a set of buffers in the Packet RAM for the receive path during initialization.
   The number of buffers is controlled by the :kconfig:option:`CONFIG_NRF700X_RX_NUM_BUFS` Kconfig option.
```

> 2. **FMAC** prepares a `CMD_RX` command and sends it to the **HAL** layer.
> 3. **HAL** sends the `CMD_RX` command to the **UMAC** firmware using the typical command flow, as described in the previous section.

Reviewer 3: ```suggestion
#. **FMAC** prepares a ``CMD_RX`` command and sends it to the **HAL** layer.
#. **HAL** sends the ``CMD_RX`` command to the **UMAC** firmware using the typical command flow, as described in the previous section.
```

Flow: Receive
-------------
The nRF70 Series Wi-Fi driver implements the following flow for receiving a frame:

> 1. **UMAC** firmware receives a frame from the air, processes it and stores it in the Packet RAM, prepares `CMD_RX` and generates an interrupt to the host processor
> 2. **FMAC** receives the interrupt, processes `CMD_RX`, copies the frame from the Packet RAM to the receive buffer.
> 3. **FMAC** performs 802.11 de-capsulation and passes the frame to the networking stack.

Reviewer 3: ```suggestion
1. **UMAC** firmware receives a frame from the air, processes it, and stores it in the Packet RAM.
   It also prepares the ``CMD_RX`` command and generates an interrupt to the host processor
#. **FMAC** receives the interrupt, processes the ``CMD_RX`` command, and copies the frame from the Packet RAM to the receive buffer.
#. **FMAC** performs 802.11 decapsulation and passes the frame to the networking stack.
```

Flow: Refill
-------------
The nRF70 Series Wi-Fi driver implements a refill mechanism to ensure that the Packet RAM is always filled with receive buffers.

> 1. **FMAC** allocates a new buffer for the received frame.
> 2. **FMAC** prepares a `CMD_RX` command and sends it to the **HAL** layer.
> 3. **HAL** sends the `CMD_RX` command to the **UMAC** firmware using the typical command flow, as described in the previous section.

Reviewer 3: ```suggestion
1. **FMAC** allocates a new buffer for the received frame.
#. **FMAC** prepares a ``CMD_RX`` command and sends it to the **HAL** layer.
#. **HAL** sends the ``CMD_RX`` command to the **UMAC** firmware using the typical command flow, as described in the previous section.
```
